File: packages/core/src/utils/hook-kind.ts
import type { HookKind } from "../models/runtime-state.js";

/**
 * R23 伏笔类型标注（393 号，四轮 P1；对标蛙趣 8 类伏笔生命周期的取精版）。
 *
 * 规范分类 ≤7 类不做数量战：承诺/悬念/危机/物品/信息/情感/世界观。
 * 职责：把 settler/architect/作者输入的自由文本归一化为规范 kind
 * （别名表精确匹配，未知 → undefined 绝不臆测）；双语展示标签。
 * 双端契约：`engine-rs/src/utils/hook_kind.rs` 1:1 镜像；共享向量
 * `golden/hook-kind-vectors.json`（TS：`golden-hook-kind.test.ts`；
 * Rust 差分：`tests/golden_hook_kind_diff.rs`）。
 */

export const HOOK_KIND_IDS = [
  "promise",
  "suspense",
  "crisis",
  "artifact",
  "information",
  "emotion",
  "worldview",
] as const;

export type HookKindId = (typeof HOOK_KIND_IDS)[number];

/** 别名表：小写化后精确匹配（zh 别名原样；en 大小写不敏感）。 */
const KIND_ALIASES: Readonly<Record<string, HookKind>> = {
  // promise 承诺
  "promise": "promise",
  "承诺": "promise",
  "约定": "promise",
  "誓言": "promise",
  // suspense 悬念
  "suspense": "suspense",
  "mystery": "suspense",
  "悬念": "suspense",
  "谜团": "suspense",
  // crisis 危机
  "crisis": "crisis",
  "危机": "crisis",
  "险局": "crisis",
  "威胁": "crisis",
  // artifact 物品
  "artifact": "artifact",
  "relic": "artifact",
  "物品": "artifact",
  "信物": "artifact",
  "道具": "artifact",
  // information 信息
  "information": "information",
  "info": "information",
  "secret": "information",
  "信息": "information",
  "情报": "information",
  "秘密": "information",
  // emotion 情感
  "emotion": "emotion",
  "relationship": "emotion",
  "romance": "emotion",
  "情感": "emotion",
  "感情": "emotion",
  "关系": "emotion",
  // worldview 世界观
  "worldview": "worldview",
  "lore": "worldview",
  "世界观": "worldview",
  "设定": "worldview",
};

/** 自由文本 → 规范 kind；未知返回 undefined（绝不臆测归类）。 */
export function normalizeHookKind(raw: string): HookKind | undefined {
  const key = raw.trim().toLowerCase();
  return KIND_ALIASES[key];
}

/** 双语展示标签（UI/timeline/回收提示）。 */
export function hookKindLabel(kind: HookKind, language: "zh" | "en" = "zh"): string {
  const zh: Record<HookKind, string> = {
    promise: "承诺",
    suspense: "悬念",
    crisis: "危机",
    artifact: "物品",
    information: "信息",
    emotion: "情感",
    worldview: "世界观",
  };
  const en: Record<HookKind, string> = {
    promise: "Promise",
    suspense: "Suspense",
    crisis: "Crisis",
    artifact: "Artifact",
    information: "Information",
    emotion: "Emotion",
    worldview: "Worldview",
  };
  return language === "en" ? en[kind] : zh[kind];
}

File: engine-rs/src/utils/hook_kind.rs
//! R23 伏笔类型标注（393 号，四轮 P1；对标蛙趣 8 类伏笔生命周期的取精版）。
//!
//! TS 真源：`packages/core/src/utils/hook-kind.ts`。规范分类 ≤7 类不做
//! 数量战：承诺/悬念/危机/物品/信息/情感/世界观。自由文本经别名表精确
//! 归一化（未知 → None 绝不臆测）；双语展示标签。
//! 共享向量 `packages/core/src/__tests__/golden/hook-kind-vectors.json`
//! 差分：`tests/golden_hook_kind_diff.rs`。

use crate::models::runtime_state::HookKind;

pub const HOOK_KIND_IDS: [HookKind; 7] = [
    HookKind::Promise,
    HookKind::Suspense,
    HookKind::Crisis,
    HookKind::Artifact,
    HookKind::Information,
    HookKind::Emotion,
    HookKind::Worldview,
];

fn kind_id(kind: HookKind) -> &'static str {
    match kind {
        HookKind::Promise => "promise",
        HookKind::Suspense => "suspense",
        HookKind::Crisis => "crisis",
        HookKind::Artifact => "artifact",
        HookKind::Information => "information",
        HookKind::Emotion => "emotion",
        HookKind::Worldview => "worldview",
    }
}

/// 别名表查询（输入已小写化/trim；zh 别名原样，en 大小写不敏感）。
fn alias_lookup(key: &str) -> Option<HookKind> {
    Some(match key {
        // promise 承诺
        "promise" | "承诺" | "约定" | "誓言" => HookKind::Promise,
        // suspense 悬念
        "suspense" | "mystery" | "悬念" | "谜团" => HookKind::Suspense,
        // crisis 危机
        "crisis" | "危机" | "险局" | "威胁" => HookKind::Crisis,
        // artifact 物品
        "artifact" | "relic" | "物品" | "信物" | "道具" => HookKind::Artifact,
        // information 信息
        "information" | "info" | "secret" | "信息" | "情报" | "秘密" => HookKind::Information,
        // emotion 情感
        "emotion" | "relationship" | "romance" | "情感" | "感情" | "关系" => HookKind::Emotion,
        // worldview 世界观
        "worldview" | "lore" | "世界观" | "设定" => HookKind::Worldview,
        _ => return None,
    })
}

/// 自由文本 → 规范 kind；未知返回 None（绝不臆测归类）。
pub fn normalize_hook_kind(raw: &str) -> Option<HookKind> {
    alias_lookup(raw.trim().to_lowercase().as_str())
}

/// 规范 id 字符串（golden 向量与序列化对齐用）。
pub fn hook_kind_id(kind: HookKind) -> &'static str {
    kind_id(kind)
}

/// 双语展示标签（UI/timeline/回收提示）。
pub fn hook_kind_label(kind: HookKind, language: &str) -> &'static str {
    match language {
        "en" => match kind {
            HookKind::Promise => "Promise",
            HookKind::Suspense => "Suspense",
            HookKind::Crisis => "Crisis",
            HookKind::Artifact => "Artifact",
            HookKind::Information => "Information",
            HookKind::Emotion => "Emotion",
            HookKind::Worldview => "Worldview",
        },
        _ => match kind {
            HookKind::Promise => "承诺",
            HookKind::Suspense => "悬念",
            HookKind::Crisis => "危机",
            HookKind::Artifact => "物品",
            HookKind::Information => "信息",
            HookKind::Emotion => "情感",
            HookKind::Worldview => "世界观",
        },
    }
}
